internal/clients/nightfall/nightfall_test.go

```diff
@@ -117,41 +116,88 @@ func (n *nightfallTestSuite) TestScan() {
-	expectedScanResp := [][]nightfallAPI.ScanResponse{
-		{},
-		{
-			{
-				Fragment: exampleCreditCardNumber,
-				Detector: string(nightfallAPI.CREDIT_CARD_NUMBER),
-				Confidence: nightfallAPI.ScanResponseConfidence{
-					Bucket: string(nightfallAPI.LIKELY),
-				},
-			},
-		},
-		{},
-	}
-
-	mockAPIClient.EXPECT().ScanAPI().Return(mockScanAPI)
-	mockScanAPI.EXPECT().ScanPayload(gomock.Any(), gomock.AssignableToTypeOf(expectedScanReq)).Return(expectedScanResp, nil, nil)
+	expectedScanReq := client.CreateScanRequest(items)
+	mockAPIClient.EXPECT().ScanPayload(gomock.Any(), expectedScanReq).Return(expectedScanResponse, nil, nil)
 
 	resp, err := client.Scan(context.Background(), githublogger.NewDefaultGithubLogger(), items)
 	n.NoError(err, "Received error from Scan")
-	n.Equal(expectedScanResp, resp, "Received incorrect response from Scan")
+	n.Equal(expectedScanResponse, resp, "Received incorrect response from Scan")
 }
 
-func createScanReq(dets []*nightfallAPI.Detector, items []string) nightfallAPI.ScanRequest {
-	detectors := make([]nightfallAPI.ScanRequestDetectors, 0, len(dets))
-	for d := range dets {
-		detectors = append(detectors, nightfallAPI.ScanRequestDetectors{
-			Name: string(d),
-		})
+func (n *nightfallTestSuite) TestScanRetries() {
+	ctrl := gomock.NewController(n.T())
+	defer ctrl.Finish()
+	cc := nightfallAPI.CREDIT_CARD_NUMBER
+	phone := nightfallAPI.PHONE_NUMBER
+	detectors := []*nightfallAPI.Detector{&cc, &phone}
+	items := []string{
+		"this is a string",
+		fmt.Sprintf("this has a credit card number %s", exampleCreditCardNumber),
+		"tom cruise!!!!!!",
 	}
-	return nightfallAPI.ScanRequest{
+	client := nightfall.Client{
 		Detectors: detectors,
-		Payload: nightfallAPI.ScanRequestPayload{
-			Items: items,
+	}
+	expectedScanReq := client.CreateScanRequest(items)
+	expectedTooManyRequestsHTTPResponse := &http.Response{StatusCode: http.StatusTooManyRequests}
+
+	tests := []struct {
+		haveSuccess        bool
+		haveNumReqAttempts int
+		wantResponse       [][]nightfallAPI.ScanResponse
+		desc               string
+	}{
+		{
+			haveSuccess:        true,
+			haveNumReqAttempts: 1,
+			wantResponse:       expectedScanResponse,
+			desc:               "success on first scan request",
 		},
+		{
+			haveSuccess:        true,
+			haveNumReqAttempts: 3,
+			wantResponse:       expectedScanResponse,
+			desc:               "success on third scan request attempt",
+		},
+		{
+			haveSuccess:        false,
+			haveNumReqAttempts: 5,
+			wantResponse:       [][]nightfallAPI.ScanResponse(nil),
+			desc:               "failed after max retries",
+		},
+	}
+	for _, tt := range tests {
+		mockAPIClient := nightfallapi_mock.NewNightfallAPI(ctrl)
+
+		client := nightfall.Client{
+			APIClient:         mockAPIClient,
+			Detectors:         detectors,
+			InitialRetryDelay: time.Millisecond,
+		}
+
+		numRetries := tt.haveNumReqAttempts
+		if tt.haveSuccess {
+			numRetries--
+		}
+		for i := 0; i < numRetries; i++ {
```

Comment:
not sure if this is possible - but can we use [.Times()](https://godoc.org/github.com/golang/mock/gomock#Call.Times) instead of this for loop? Seems like it would simplify this code a bit and allow us to put the `wantNumRetries` value into the test structs instead of calculating it on the fly like on lines 178-181.

What do you think?


Separately, since there's so much conditional logic to check the validity of each test in this test table (and not a ton of shared code), is it cleaner to use subtests? Curious for your thoughts here too.